### PR TITLE
Update to the latest YCM version + some improvements

### PR DESCRIPTION
**Pull request by drdanz**
* Update YCM files to latest version (0.8.1)
* Use EXPORT instead of TARGETS_PROPERTY in install_basic_package_files
* Require CMake 3.5
* Build position independent code and disable compiler extensions
* Add optional dependency on YCM

### CODE REVIEW

**Comment by traversaro** [comment]
To be honest, I am not a big fan of the optional dependency on YCM. The relation with YCM and the reason why those files are vendored is clearly explained in the README, and adding the optional dependency on YCM would just make it more difficult to debug the CMake build  script in case some strange version of YCM is installed in the system, without removing the problem that the internal files are  vendored and so they need to manually updated.

**Comment by drdanz** [comment]
@traversaro would a comment instead be ok for you?
```cmake
# If you don't want to ship these files with your code (recommended), you can
# instead depend on YCM.
# In this case replace the following line with
# find_package(YCM 0.8.1 REQUIRED)
list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake)
```

**Comment by drdanz** [comment]
Replaced optional dependency on YCM with a comment.
I also added a new commit, fixing the `RPATH` support (`INSTALL_NAME_DIR` should point to the libraries are installed, i.e. `CMAKE_INSTALL_FULL_LIBDIR`, not to `CMAKE_INSTALL_PREFIX`), and removing the `ENABLE_RPATH` option in favour of CMake standard options (fixes #7).

File: CMakeLists.txt
Comment (claudiofantacci) [inline CMakeLists.txt]: Since this project is also a tutorial, I'd like to give room to comments to explain why we're setting this feature. For example, we could add something like:

> Position Independent Code (PIC) is commonly used for shared libraries so that the same shared library code can be loaded in each program address space in a location where it will not overlap with any other uses of such memory. In particular, this option avoids problems occurring when a process wants to load more than one shared library at the same virtual address. Since shared libraries cannot predict where other shared libraries could be loaded, this is an unavoidable problem with the traditional shared library concept. Generating position-independent code is often the default behavior for most modern compilers.
Further details on PIC can be found here: https://eli.thegreenplace.net/2011/11/03/position-independent-code-pic-in-shared-libraries/


which is elaborated from Wikipedia. If you have precise hints about this command please feel free to add or modify the description.

File: CMakeLists.txt
Comment (claudiofantacci) [inline CMakeLists.txt]: Same as above 😄 . I would add something like

> C/CXX_EXTENSIONS are ON by default to allow the compilers to use extended variants of the C/CXX language. However, this could expose cross-platform bugs in user code or in the headers of third-party dependencies and thus it is strongly suggested to turn extensions off.

which is an elaborated version from CMake documentation itself.

File: CMakeLists.txt
Comment (claudiofantacci) [inline CMakeLists.txt]: Awesome, thanks!
I could not be able to find the `CMAKE_INSTALL_FULL_LIBDIR/BINDIR` is it coming as an utility variables from somewhere? If so, could you please add a comment about it specifying what these variables contain.

File: src/LibTemplateCMake/CMakeLists.txt
Comment (claudiofantacci) [inline src/LibTemplateCMake/CMakeLists.txt:10]: Thanks for this nice comment. I'm wondering whether it would be more suitable for a tutorial to explicitly use `FIRST_TARGET` (which I agree would not be necessary here, but is for the sake of teaching 😄) and then write that in this case it would have been sufficient to use the same `PROJECT_NAME` without `FIRST_TARGET`.
Either way, I'm fine so it is up to you to choose 😄

**Review by claudiofantacci — CHANGES_REQUESTED** [review]
Thanks for the contribution!
I just wrote down a copule of comments of which I would like to know your opinion.

**Comment by claudiofantacci** [comment]
Please also note that @mbrunettini is doing some tests with Genkins on master branch, so please do not merge this until he finishes.

Please @mbrunettini let us know when you are ok with the test so that we can clean master commits and merge this PR!

Thanks everyone for the collaboration 🎉

File: CMakeLists.txt
Comment (drdanz) [reply CMakeLists.txt]: The `CMAKE_INSTALL_FULL_<dir>` variables are well documented in the [GNUInstallDirs documentation](https://cmake.org/cmake/help/git-master/module/GNUInstallDirs.html) I don't think any explaination here is necessary

Comment (claudiofantacci) [reply CMakeLists.txt]: [My tirivial search criteria got me!](https://cmake.org/cmake/help/latest/search.html?q=CMAKE_INSTALL_FULL_LIBDIR&check_keywords=yes&area=default)
Thanks for pointing to GNUInstallDirs documentation 😄

File: src/LibTemplateCMake/CMakeLists.txt
Comment (drdanz) [reply src/LibTemplateCMake/CMakeLists.txt:10]: I considered that as well, but I think in this case it is not really useful, it does not teach you anything more than what is explained in the comment. Also if you need to do something more complicated, you can easily read the documentation of the modules...

**Comment by drdanz** [comment]
Updated with the comments requested (I added an extra comment on the PIC one about shared libraries linking non-PIC static ones) and added links to YCM modules documentation

**Review by claudiofantacci — DISMISSED** [review]
Great! Thanks for changes and the discussion.

**Comment by claudiofantacci** [comment]
Please don't merge until @mbrunettini has finished with Jenkins tests.

File: CMakeLists.txt
Comment (traversaro) [inline CMakeLists.txt:108]: What's the option to enable/disable install rpath equivalent to `ENABLE_RPATH` ? Even if it is a CMake variable, should we explicitly list as an option as we do for `BUILD_SHARED_LIBS` ?

Comment (claudiofantacci) [reply CMakeLists.txt:108]: It is `CMAKE_SKIP_RPATH` which automatically pops out at configuration time.
If we make it as an option it would probably be redundand, isn't it @drdanz ?

**Comment by claudiofantacci** [comment]
@drdanz can you please rebase on master and remove the commits related to the Jenkins files? After that I merge the PR. Thanks!

**Comment by drdanz** [comment]
@claudiofantacci done.

**Comment by claudiofantacci** [comment]
Thanks @drdanz!

**Pull request merged**
